python/tank/descriptor/bundle_cache_usage/manager.py

```diff
@@ -0,0 +1,270 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import time
+import threading
+
+from worker import BundleCacheUsageWorker
+from exception import BundleCacheUsageException
+from exception import BundleCacheUsageFileDeletionException
+from . import BundleCacheUsageLogger as log
+
+
+class BundleCacheManager(object):
```

Comment:
we have a singleton base class. can we use that?
https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/util/singleton.py

Reply:
From experience I do prefer a double-lock style (see https://en.wikipedia.org/wiki/Double-checked_locking), which I used in my version and didn't want to impose a changes in the toolkit one. Several of my tests would fail using the tookit one. I agree, that I am stressing more than typical use. I do know that what I got is robust. Still, I'll undo that if you really want me to.